3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/OpenCageEnrichment.scala

```diff
@@ -0,0 +1,223 @@
+    isParseable(config, schemaKey).flatMap { _ =>
+      {
+        (for {
+          apiKey    <- ScalazJson4sUtils.extract[String](config, "parameters", "apiKey")
+          cacheSize <- ScalazJson4sUtils.extract[Int](config, "parameters", "cacheSize")
+          scale     <- ScalazJson4sUtils.extract[Int](config, "parameters", "scale")
+          timeout   <- ScalazJson4sUtils.extract[Int](config, "parameters", "timeout")
+        } yield OpenCageEnrichment(apiKey, cacheSize, scale, timeout)).toValidationNel
+      }
+    }
+}
+
+/**
+ * OpenCage API makes it possible to convert coordinates to and from places (e.g. 51.4266,-0.0798 ->
+ * Shoreditch, London)
+ *
+ * @param apiKey OpenCage's API key
+ * @param cacheSize amount of locations with prefetched geolocation info
+ * @param scale Number of decimals to keep from coordinates. Above 6 is useless
+ * @param timeout Number of seconds to wait for OpenCage API to respond
+ */
+case class OpenCageEnrichment(apiKey: String, cacheSize: Int, scale: Int, timeout: Int) extends Enrichment {
+
+  type Geocode = parts.Result
+
+  private lazy val client = new OpenCageClient(apiKey)
+
+  private val schemaUri = "iglu:com.opencagedata/geocoder/jsonschema/1-0-0"
+
+  private implicit val formats: DefaultFormats.type = DefaultFormats
+
+  protected lazy val cache = OpenCageCache(client, cacheSize, scale, timeout)
+
+  /**
+   * Get textual description of a location for a specific event
+   * Any non-fatal error will return failure and thus whole event will be
+   * filtered out in future
+   *
+   * @param latitude enriched event optional latitude (probably null)
+   * @param longitude enriched event optional longitude (probably null)
+   * @return opencage location as self-describing JSON object
+   */
+  // It accepts Java Float (JFloat) instead of Scala's because it will throw NullPointerException
+  // on conversion step if `EnrichedEvent` has nulls as geo_latitude or geo_longitude
+  def getGeoCodingContext(latitude: Option[JFloat], longitude: Option[JFloat]): Validation[String, JObject] =
+    try {
+      getGeocoding(latitude, longitude).map(addSchema)
+    } catch {
+      case NonFatal(ex) => s"Failed requesting lat: $latitude and lon: $longitude with exception ${ex.toString}".fail
+    }
+
+  /**
+   * Get geocoding of a latitude, longitude pair
+   */
+  private def getGeocoding(latitude: Option[JFloat], longitude: Option[JFloat]): Validation[String, JObject] =
+    (latitude, longitude) match {
+      case (Some(lat), Some(lon)) =>
+        for {
+          response <- getCachedOrRequest(lat, lon)
+          geocode  <- getGeocodeFromResponse(response)
+          obj      <- geocodeToJValue(geocode)
+        } yield obj
+
+      case _ => s"One of the required event fields missing. latitude $latitude, longitude: $longitude".fail
+    }
+
+  private def geocodeToJValue(location: Geocode): Validation[String, JObject] =
+    Extraction.decompose(location) match {
```

Comment:
is this exception-safe?

Reply:
It is converting a known type into a Json, so I guess there is no risk here.